clib/helpstr.c

```diff
@@ -26,13 +27,92 @@
 /*  any type of equipment. An m is for Mark III, and a v is for   */
 /*  VLBA.                                                         */
 
+#include <dirent.h>
+#include <limits.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
+#include <sys/types.h>
+#include <unistd.h>
+
 #include "../include/params.h"
 
-#define MAX_STRING  256
+#define MAX_STRING 256
+
+// fileexists check if file exists and is readable
+static int fileexists(const char *filename) {
+	FILE *file;
+	if ((file = fopen(filename, "r"))) {
+		fclose(file);
+		return 1;
+	}
+	return 0;
+}
+
+static char *help_dirs[] = {FS_ROOT "/st/help", FS_ROOT "/fs/help", NULL};
+
+// check_help_dirs looks for file "name" in help dirs and returns first match as a
+// string.  Resulting string must be released with free
+static char *check_help_dirs(char *name) {
+	char **dir;
+	char *path;
+
+	for (dir = help_dirs; *dir != NULL; dir++) {
+		if (asprintf(&path, "%s/%s", *dir, name) < 0)
+			return NULL;
+		if (fileexists(path)) {
+			return path;
+		}
+		free(path);
+	}
+	return NULL;
+}
+
+// check_help_dirs_prefix returns a null terminated list of files that start with
+// prefix. Returns NULL on error.
+// List and its elements must be released by free.
+static char **ls_prefix(char **dirs, char *prefix) {
+
+	size_t prefix_len = strlen(prefix);
+	if (!prefix_len)
+		return NULL;
+
+	// length of allocated buffer
+	size_t len = 10;
+	char **ret = calloc(len + 1, sizeof(char *));
+	if (!ret)
+		return NULL;
+
+	DIR *dp;
+	struct dirent *ep;
+
+	size_t nmatches = 0;
+	char **d;
+	for (d = dirs; *d != NULL; d++) {
+		dp = opendir(*d);
+		if (dp == NULL)
+			continue;
+		while ((ep = readdir(dp))) {
+			if (strncmp(prefix, ep->d_name, prefix_len) != 0) {
+				continue;
+			}
+
+			nmatches++;
+
+			if (nmatches > len) {
+				len += len;
+				ret = realloc(ret, len + 1);
```

Comment:
Should this be something like:
````
-                              ret = realloc(ret, len + 1);
+                              ret = realloc(ret, (len + 1) * sizeof(char *));
+                              int i;
+                              for (i=len/2;i<=len;i++)
+                                  *ret[i]=0;
 ````

Reply:
Or even

    -    *ret[i]=0;
    +    ret[i]=0;

This programming business is hard.